### PR TITLE
[Merged by Bors] - feat: Dyck words up to semilength

### PR DESCRIPTION
**Pull request by Parcly-Taxel**
First part of #9781.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Combinatorics/Enumerative/DyckWord.lean [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+* `DyckWord.semilength`: semilength (half the length) of a Dyck word.
+
+## Implementation notes
+
+While any two-valued type could have been used for `DyckStep`, a new enumerated type is used here
+to emphasise that the definition of a Dyck word does not depend on that underlying type.
+
+## TODO
+
+* Prove the bijection between Dyck words and Catalan numbers.
```
Comment (kbuzzard) [inline Mathlib/Combinatorics/Enumerative/DyckWord.lean:31]: A Catalan number is a number, so "bijection" here is probably not the right word.

Comment (Parcly-Taxel) [reply Mathlib/Combinatorics/Enumerative/DyckWord.lean:31]: Right...

Comment (Parcly-Taxel) [inline Mathlib/Combinatorics/Enumerative/DyckWord.lean:31]:
```suggestion
* Prove the bijection between Dyck words and rooted binary trees.
```

File: Mathlib/Combinatorics/Enumerative/DyckWord.lean [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+lemma dyckStep_cases (s : DyckStep) : s = U ∨ s = D := by cases' s <;> tauto
+
+/-- A Dyck word is a list of `DyckStep`s with as many `U`s as `D`s and with every prefix having
+at least as many `U`s as `D`s. -/
+@[ext]
+structure DyckWord where
+  /-- The underlying list -/
+  l : List DyckStep
+  /-- There are as many `U`s as `D`s -/
+  bal : l.count U = l.count D
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/Enumerative/DyckWord.lean:53]: Letters are cheap in field names
```suggestion
  count_U_eq_count_D : l.count U = l.count D
```

File: Mathlib/Combinatorics/Enumerative/DyckWord.lean [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+  /-- Each prefix has as least as many `U`s as `D`s -/
+  nonneg i : (l.take i).count D ≤ (l.take i).count U
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/Enumerative/DyckWord.lean:55]:
```suggestion
  count_D_le_count_U i : (l.take i).count D ≤ (l.take i).count U
```

File: Mathlib/Combinatorics/Enumerative/DyckWord.lean [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+open DyckStep
+
+lemma dyckStep_cases (s : DyckStep) : s = U ∨ s = D := by cases' s <;> tauto
+
+/-- A Dyck word is a list of `DyckStep`s with as many `U`s as `D`s and with every prefix having
+at least as many `U`s as `D`s. -/
+@[ext]
+structure DyckWord where
+  /-- The underlying list -/
+  l : List DyckStep
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/Enumerative/DyckWord.lean:51]:
```suggestion
  toList : List DyckStep
```

File: Mathlib/Combinatorics/Enumerative/DyckWord.lean [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+  add_zero p := by ext1; exact append_right_eq_self.mpr rfl
+  zero_add p := by ext1; rfl
+  add_assoc p q r := by ext1; apply append_assoc
+  nsmul := nsmulRec
+
+namespace DyckWord
+
+variable (p q : DyckWord)
+
+lemma eq_zero_iff : p = 0 ↔ p.l = [] := by rw [DyckWord.ext_iff]; rfl
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/Enumerative/DyckWord.lean:76]:
```suggestion
lemma toList_eq_nil : p.toList = [] ↔ p = 0 := by rw [DyckWord.ext_iff]; rfl
```

File: Mathlib/Combinatorics/Enumerative/DyckWord.lean [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+
+lemma denest_nest (h : p.l ≠ []) (pos : ∀ i, 0 < i → i < p.l.length →
+    (p.l.take i).count D < (p.l.take i).count U) : (p.denest h pos).nest = p := by
+  simpa [DyckWord.ext_iff] using p.cons_tail_dropLast_concat h
+
+section Semilength
+
+/-- The semilength of a Dyck word is half of the number of `DyckStep`s in it, or equivalently
+its number of `U`s. -/
+def semilength : ℕ := p.l.count U
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/Enumerative/DyckWord.lean:175]: Can we have a lemma showing it’s equal to the count of D?

**Review by YaelDillies — COMMENTED** [review]
I need someone who knows more than me to confirm those are the names used in the literature for those operations. It would also help if you gave references naming those operations explicitly

File: Mathlib/Combinatorics/Enumerative/DyckWord.lean (hunk 6 above)
Comment (Parcly-Taxel) [reply Mathlib/Combinatorics/Enumerative/DyckWord.lean]: All done.

File: Mathlib/Combinatorics/Enumerative/DyckWord.lean [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+
+/-- A `DyckStep` is either `U` or `D`, corresponding to `(` and `)` respectively. -/
+inductive DyckStep
+  | U : DyckStep
+  | D : DyckStep
+  deriving Inhabited, DecidableEq
+
+open DyckStep
+
+lemma dyckStep_cases (s : DyckStep) : s = U ∨ s = D := by cases' s <;> tauto
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/Enumerative/DyckWord.lean:44]: `bool_cases` does not exist. Mind aligning your name to the existing one for `Bool`?

File: Mathlib/Combinatorics/Enumerative/DyckWord.lean [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+  add p q := ⟨p ++ q, by
+    simp only [count_append, p.count_U_eq_count_D, q.count_U_eq_count_D], by
+    simp only [take_append_eq_append_take, count_append]
+    exact fun _ ↦ add_le_add (p.count_D_le_count_U _) (q.count_D_le_count_U _)⟩
+
+instance : Zero DyckWord := ⟨[], by simp, by simp⟩
+
+/-- Dyck words form an additive monoid under concatenation, with the empty word as 0. -/
+instance : AddMonoid DyckWord where
+  add_zero p := by ext1; exact append_right_eq_self.mpr rfl
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/Enumerative/DyckWord.lean:71]:
```suggestion
  add_zero p := by ext1; exact append_nil _
```

Comment (Parcly-Taxel) [reply Mathlib/Combinatorics/Enumerative/DyckWord.lean]: Done.

**Review by YaelDillies — APPROVED** [review]
Looking good now. Thanks!

maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**